agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
-        );
-        Err(Status::new(
-            Code::Unknown,
-            "Could not find device usage slot",
-        ))
+        Err(_) => (0..*capacity)
+            .map(|x| {
+                (
+                    format!("{}-{}", instance_name, x),
+                    DeviceUsageStatus::Unknown,
+                )
+            })
+            .collect(),
     }
 }
 
-/// This tries up to `MAX_INSTANCE_UPDATE_TRIES` to update the requested slot of the Instance with the this node's name.
-/// It cannot be assumed that this will successfully update Instance on first try since Device Plugins on other nodes may be simultaneously trying to update the Instance.
-/// This returns an error if slot does not need to be updated or `MAX_INSTANCE_UPDATE_TRIES` attempted.
-async fn try_update_instance_device_usage(
-    device_usage_id: &str,
+async fn get_available_virtual_devices(
+    instance_config: &InstanceConfig,
     node_name: &str,
-    instance_name: &str,
     instance_namespace: &str,
+    capacity: &i32,
     kube_interface: Arc<impl KubeInterface>,
-) -> Result<(), Status> {
-    let mut instance: InstanceSpec;
-    for x in 0..MAX_INSTANCE_UPDATE_TRIES {
-        // Grab latest instance
-        match kube_interface
-            .find_instance(instance_name, instance_namespace)
-            .await
-        {
-            Ok(instance_object) => instance = instance_object.spec,
-            Err(_) => {
-                trace!(
-                    "try_update_instance_device_usage - could not find Instance {}",
-                    instance_name
-                );
-                return Err(Status::new(
-                    Code::Unknown,
+) -> HashSet<String> {
+    let mut device_usage_states = HashMap::new();
+    for instance_name in instance_config.instances.keys() {
+        let device_usage_state = get_instance_device_usage_state(
+            node_name,
+            instance_name,
+            instance_namespace,
+            capacity,
+            kube_interface.clone(),
+        )
+        .await;
+        device_usage_states.insert(instance_name.to_string(), device_usage_state);
+    }
+    let mut vdev_ids_to_report = HashSet::new();
+    let mut free_instances = HashSet::new();
+    let mut reserved_instances = HashSet::new();
+    for (instance_name, device_usage_state) in device_usage_states {
+        for (_id, state) in device_usage_state {
+            match state {
+                DeviceUsageStatus::Free => {
+                    free_instances.insert(instance_name.clone());
+                }
+                DeviceUsageStatus::ReservedByConfiguration(vdev_id) => {
+                    reserved_instances.insert(instance_name.clone());
+                    vdev_ids_to_report.insert(vdev_id);
+                }
+                _ => (),
+            };
+        }
+    }
+
+    // Decide virtual device id, report already reserved virtual devices + 1 free device usage (if available) per instance
```

Comment:
why are `already reserved virtual devices` returned from a function named get_available_virtual_devices?  are they available?  should this function be renamed to something like get_virtual_devices_not_used_by_other_nodes?  or maybe that is just what "available" means in this context.  either way, more comments would help.

Reply:
maybe available=healthy?

can this function include this and return healthy devices to simplify ConfigurationDevicePlugin.list_and_watch?

```
            let current_virtual_devices = available_virtual_devices
                .into_iter()
                .map(|id| (id, HEALTHY.to_string()))
                .collect::<HashMap<String, String>>();
```

Reply:
I prefer to keep get_available_virtual_devices() neutral and have list_and_watch handle the returned result so the function get_available_virtual_devices() can be reused

Reply:
makes sense, the reason i suggested it is that the function is not reused.

Reply:
Single responsibility principle :)

Reply:
lol ... indeed.  unless the function is renamed to get_healthy_devices